### PR TITLE
add soft-reboot

### PR DESCRIPTION
**Pull request by YukariChiba**
this commit adds initial support for `soft-reboot` command to `shutdown`, which is similar to [systemd's soft-reboot](https://www.freedesktop.org/software/systemd/man/latest/systemd-soft-reboot.service.html), but without root switching.

Early services need to be idempotent for soft-reboot to work properly.

---

~~Updated: added an argument `--ignore-sys-args` to ignore kernel args, which would be added automatically in soft-reboot `execv`.~~

~~Updated: filtered out "single" argument~~

Updated: The re-`execv`-ing retains boot-time parameters to simplify the code and is consistent with systemd's soft-reboot behavior.

### CODE REVIEW

File: src/dinit.cc
Comment (davmac314) [reply src/dinit.cc:705]: (I just noticed this PR is still marked as WIP. Apologies if you hadn't finished yet - I will wait until you remove the WIP status before I review any further, but in general it looks ok otherwise).

Comment (YukariChiba) [reply src/dinit.cc:705]: I thought of this and have several possible ideas.

1. strip the `single` argument from the `argv`, which works, but the strip process is complicated and requires more modification if new acceptable kernel arguments are added subsequently.
2. set a new cmdline argument `--no-process-sys-args`, and place it at the beginning of the `argv` when re-`execv`-ing.
3. add a new argument to `shutdown` that overrides the `argv` when re-`execv`-ing.

I think this needs further discussion.

Comment (YukariChiba) [reply src/dinit.cc]: Added an cmdline arg `--ignore-sys-args` to ignore kernel args like `single`

File: src/dinit.cc
Comment (mobin-2008) [inline src/dinit.cc:711]: One of the main principles of Dinit is ["Robustness (including allocation robustness)"](https://github.com/davmac314/dinit/blob/0b5158328427778f1e7d28104bc8cca43aef572a/doc/COMPARISON#L27-L31). The `malloc()` call could fail and it's not handled here and can result in null dereference in further usage (see L713).

I think malloc failure should be handled here.

Comment (davmac314) [reply src/dinit.cc:711]: I need a little time to consider the whole issue. I think it is probably better to filter arguments out before re-exec, or even to re-build a new command line from scratch, than it is to have a new argument which effectively filters out arguments at a different stage. Please allow me a little time to get back to this issue with a response once I've thought it through. @mobin-2008 is correct though that we cannot ignore `malloc` failures. It's also generally preferable to use `new` rather than `malloc`, note that this throws an exception on failure and that needs to be handled.

Comment (YukariChiba) [reply src/dinit.cc]: Okay, I've now filtered the `single` argument in `argv`, and handled the exception in `new`.

File: src/dinit.cc
Comment (mobin-2008) [inline src/dinit.cc:720]: As far as I know, a object created with `new[]` should be freed with `delete[]` not `free()`.
```suggestion
                delete[] argv_exec;
```

File: src/dinit.cc
Comment (mobin-2008) [inline src/dinit.cc:721]: From [Dinit's code style guide](https://github.com/davmac314/dinit/blob/master/doc/CODE-STYLE):
> 2. Curly braces - for classes and functions, the opening brace appears by itself on an empty line, as does the closing brace. In other cases the opening brace should appear at the end of the line (and not on a line by itself). The closing brace always appears on a line by itself.
```suggestion
            }
            catch (std::bad_alloc &exc) {
```

File: src/dinit.cc
Comment (davmac314) [inline src/dinit.cc:711]: Sorry for not being able to look at this earlier. As I mentioned, I needed to find a time when I wasn't busy to figure out the right solution. Having thought about it for a bit, I think I have some better idea now.

This current solution, filtering just the "single" argument, isn't sufficient because there are a number of ways users can pass additional service names on the command line argument (via the bootloader):

- `single` (on Linux) which is covered by your current changes
- `-t servicename` for any service name or `--service servicename` for any service name
- `-m servicename1 servicename2 servicename...` for any service names
- `-c servicename1 servicename2 servicename...` for any service names (container mode)

I think the best option is to, when processing the command line arguments initially, construct an array (probably a `vector<boolean>` actually) which keeps track of which arguments should be filtered out on re-exec. Then we don't need to duplicate the argument processing logic; it's all in one place.

File: src/dinit.cc
Comment (davmac314) [inline src/dinit.cc:718]: If the `execv` actually fails, it would be good to log why (making use of `strerror` to convert the errno into a message).

**Review by davmac314 — CHANGES_REQUESTED** [review]

File: src/dinit.cc
Comment (davmac314) [reply src/dinit.cc:711]: Alternatively, update the documentation to be very clear about what a "soft reboot" actually does. It seems reasonable to me that it will just restart with the exact same arguments, but in that case the documentation needs to be clear and warn that the original boot-time options will be re-applied.

Comment (YukariChiba) [reply src/dinit.cc]: I just tested this on systemd and found that the `single` parameter is still preserved after soft-reboot, so let's just use `argv` directly to be consistent with it in this behavior.

Comment (YukariChiba) [reply src/dinit.cc]: > Usecase: this is an extremely quick way to reset userspace fully when updating image based systems, without going through a full hardware/firmware/boot loader/kernel reset. It minimizes "grayout time" for OS updates. (In particular when combined with kernel live patching)

From the original PR for systemd: https://github.com/systemd/systemd/pull/27435

**Review by davmac314 — CHANGES_REQUESTED** [review]
Otherwise looks mostly ok, although you haven't adjusted the documentation to make the precise action of a soft-shutdown more explicitly documented. I can do that myself after merging.

**Comment by YukariChiba** [comment]
The behavior of soft-reboot has been described in the documentation.

**Review by davmac314 — APPROVED** [review]

**Comment by davmac314** [comment]
Thanks, now merged. If you would like to add yourself to the CONTRIBUTORS file, please open another pull request to do so. :)

**Pull request merged**
